app/models/abstract_feature.rb

```diff
@@ -117,15 +117,54 @@ def self.cache_derivatives(options = {})
     SQL
 
     update_all <<-SQL.squish
-      geom_lowres  = ST_SimplifyPreserveTopology(geom, #{options.fetch(:lowres_simplification, lowres_simplification)})
+      geom_lowres  = ST_SimplifyPreserveTopology(geom, #{options.fetch(:lowres_simplification, lowres_simplification)}),
+      tilegeom     = ST_Transform(geom, 3857)
     SQL
 
     invalid('geom_lowres').update_all <<-SQL.squish
       geom_lowres  = ST_Buffer(geom_lowres, 0)
     SQL
   end
 
-  def self.geojson(lowres: false, precision: 6, properties: true, srid: 4326, centroids: false) # default srid is 4326 so output is Google Maps compatible
+  def self.mvt(*args)
+    select_sql = mvt_sql(*args)
+
+    # Result is a hex string representing the desired binary output so we need to convert it to binary
+    result = SpatialFeatures::Utils.select_db_value(select_sql)
+    result.remove!(/^\\x/)
+    result = result.scan(/../).map(&:hex).pack('c*')
+
+    return result
+  end
+
+  def self.mvt_sql(tile_x, tile_y, zoom, properties: true, centroids: false, metadata: {}, scope: nil)
+    if centroids
+      column = 'ST_Transform(centroid::geometry, 3857)' # MVT works in SRID 3857
+    else
+      column = 'tilegeom'
+    end
+
+    subquery = all
+    subquery = subquery
+                .select("ST_AsMVTGeom(#{column}, ST_TileEnvelope(#{zoom}, #{tile_x}, #{tile_y}), extent => 4096, buffer => 64) AS geom")
+                .select('id')
+                .where("#{column} && ST_TileEnvelope(#{zoom}, #{tile_x}, #{tile_y}, margin => (64.0 / 4096))")
```

Comment:
Could we just do:

```suggestion
    subquery = select("ST_AsMVTGeom(#{column}, ST_TileEnvelope(#{zoom}, #{tile_x}, #{tile_y}), extent => 4096, buffer => 64) AS geom")
                .select('id')
                .where("#{column} && ST_TileEnvelope(#{zoom}, #{tile_x}, #{tile_y}, margin => (64.0 / 4096))")
```

Reply:
Yeah subquery was in there as a cut-and-paste from the geojson method at one point, but both had the scoping refactored out.